Customization of Xcode build configuration names

danieldickison opened the pull request:
Patch for issue #588.

Adds two new configs, `podfileDebugConfigurations` and `podfileReleaseConfigurations`, to allow customization of the Xcode build configuration names that are specified in the Podfile as linking with the debug and release translated libs. These default to the previously hardcoded `['Debug']` and `['Release']`.

brunobowden commented on FAQ.md:
````diff
@@ -336,6 +337,24 @@ you'd like to access from Swift code.
 ```
 
 
+### How do I specify additional Xcode build configurations?
````
It's a podfile configuration rather than Xcode, so title it:

    How do I specify additional Podfile build configurations for Xcode?

Then also update the title and link in the TOC.

brunobowden replied:
Actually, this is a configuration inside Xcode as well? I'm wondering if this should be named xcodeDebugConfigurations and xcodeReleaseConfigurations instead? Please comment on this.

brunobowden commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy:
```diff
@@ -101,11 +107,16 @@ class XcodeTask extends DefaultTask {
 
-        PodspecDetails(String projectNameIn, File podspecDebugIn, File podspecReleaseIn) {
+        PodspecDetails(String projectNameIn, File podspecDebugIn, File podspecReleaseIn,
+                       List<String> debugConfigurationsIn, List<String> releaseConfigurationsIn) {
```
Use variable name `podfileReleaseConfigurations`... I like to reuse the name so a single search finds all the relevant code and comments, rather than having to follow the chain of logic.

danieldickison commented:
Not sure what the best name for this would be: the build configurations are things defined in the Xcode project, but the Podfile references those configuration names. I guess that's the same situation as target names and those configs are named e.g. `xcodeTargetsIos`, so I can rename these to be `xcodeDebugConfigurations` and `xcodeReleaseConfigurations` if you prefer.

brunobowden commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy:
```diff
@@ -113,6 +124,14 @@ class XcodeTask extends DefaultTask {
         }
 
+        String getJoinedDebugConfigurations() {
+            return debugConfigurations.collect {"'$it'"}.join(', ')
```
Again reuse the `podfileDebugConfigurations` so a single search finds all the references. You can append or prepend that variable name as needed.

Extract this to a utility method in Utils.groovy and add an appropriate unit test:

    def toQuotedList(List<String> listString) {
        return "'${listString.toArray().join("','")}'"
    }

This does the magic that you're looking for ;-) If you're doing stuff with iterators in the future you need to add explicit types. See the style guidelines on this.

https://github.com/j2objc-contrib/j2objc-gradle/blob/master/CONTRIBUTING.md#style-guidelines

brunobowden replied:
Then it can be:

    podMethodLines.add("    pod '$podspecDebugName', :configuration => [${Utils.toQuotedList(xcodeDebugConfigurations))}], :path => '$pathDebug'".toString())

danieldickison replied:
Is there a reason for the `toArray()` there, as opposed to calling join on the List directly? I.e.

    return "'#{listString.join("','")}'"

brunobowden commented on src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTaskTest.groovy:
```diff
@@ -1107,6 +1112,27 @@ class XcodeTaskTest {
+        List<String> expectedLines = [
+                "def j2objc_PROJ",
+                "    pod 'j2objc-PROJ-debug', :configuration => ['Debug', 'Beta'], :path => '../PROJ/BUILD'",
+                "    pod 'j2objc-PROJ-release', :configuration => ['Release', 'Preview'], :path => '../PROJ/BUILD'",
```
Good, this is the test you need to have.

brunobowden commented:
More important than what I prefer is what makes best sense for other engineers. I had mistakenly thought this was only for the Podfile... but since this is an Xcode configuration then it feel like it should be consistent with the other settings. So xcodeDebugConfigurations and xcodeReleaseConfigurations sounds like a better choice. Feel free to push back if you feel differently.

If you're good with this, update the methods, tests and FAQ and it should be good for a final review.

danieldickison commented:
Yup I agree the xcode* naming makes more sense now that I've seen more of the context. I'll make those changes and update the pull request.

I'm now getting a bizarre build error from groovy on identical code that previously compiled :/ I'll update when I get this sorted, but I'm curious if you've seen something like this.

```* What went wrong:
Execution failed for task ':compileGroovy'.
> BUG! exception in phase 'instruction selection' in source unit '/Users/daniel/bc/j2objc-gradle/src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy' Declaring class for method call to 'java.util.List getXcodeDebugConfigurations()' declared in com.github.j2objccontrib.j2objcgradle.tasks.XcodeTask was not matched with found receiver com.github.j2objccontrib.j2objcgradle.tasks.XcodeTask$PodspecDetails. This should not have happened!
```

brunobowden commented:
I didn't think it worked without `toArray()` but bonus points if you
successfully make it simpler. Try clean before another build.

Ping me when it's ready and I'll do a final review.

On Wed, Jan 20, 2016 at 3:18 PM Daniel Dickison <notifications@github.com>
wrote:

> Yup I agree the xcode* naming makes more sense now that I've seen more of
> the context. I'll make those changes and update the pull request.
>
> I'm now getting a bizarre build error from groovy on identical code that
> previously compiled :/ I'll update when I get this sorted, but I'm curious
> if you've seen something like this.
>
> ```* What went wrong:
> Execution failed for task ':compileGroovy'.
>
> BUG! exception in phase 'instruction selection' in source unit
> '/Users/daniel/bc/j2objc-gradle/src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy'
> Declaring class for method call to 'java.util.List
> getXcodeDebugConfigurations()' declared in
> com.github.j2objccontrib.j2objcgradle.tasks.XcodeTask was not matched with
> found receiver
> com.github.j2objccontrib.j2objcgradle.tasks.XcodeTask$PodspecDetails. This
> should not have happened!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/j2objc-contrib/j2objc-gradle/pull/590#issuecomment-173397111>
> .
>

danieldickison commented:
@brunobowden that's ready for review.

As mentioned in the commit message, I couldn't get around the strange groovy error so I left a workaround in, renaming the offending methods with a "2" suffix. In theory those methods could be renamed back without the "2" suffix, but I'm unable to get that to compile locally on my machine even after clean and wiping out gradle cache directories (I guess there must be a groovy compiler cache directory I'm missing since it persists after rebooting).

brunobowden commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy:
```diff
@@ -87,6 +87,14 @@ class XcodeTask extends DefaultTask {
     @Input
     String getMinVersionWatchos() { return J2objcConfig.from(project).minVersionWatchos }
 
+    // The '2' suffix here is purely to get around what appears to be a groovy compiler bug.
```
Make this a TODO to revert to the proper name without '2'

brunobowden commented on FAQ.md:
````diff
@@ -336,6 +337,24 @@ you'd like to access from Swift code.
+```gradle
+// File: shared/build.gradle
+j2objcConfig {
+    xcodeReleaseConfigurations += ['Preview']
````
If you think it fits, have both configs:

    xcodeDebugConfigurations += ['Beta']
    xcodeReleaseConfigurations += ['Preview']

brunobowden commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTask.groovy:
```diff
@@ -666,8 +680,8 @@ class XcodeTask extends DefaultTask {
         podMethodLines.add("def ${podspecDetails.getPodMethodName()}".toString())
-        podMethodLines.add("    pod '$podspecDebugName', :configuration => ['Debug'], :path => '$pathDebug'".toString())
-        podMethodLines.add("    pod '$podspecReleaseName', :configuration => ['Release'], :path => '$pathRelease'".toString())
+        podMethodLines.add("    pod '$podspecDebugName', :configuration => [${Utils.toQuotedList(podspecDetails.xcodeDebugConfigurations)}], :path => '$pathDebug'".toString())
```
These lines are getting along, so extract it:

    String xcodeDebugConfigs = Utils.toQuotedList(podspecDetails.xcodeDebugConfigurations)

Then in the string:

    "    pod '$podspecDebugName', :configuration => [$xcodeDebugConfigs], :path => '$pathDebug'".toString()

brunobowden commented on src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/UtilsTest.groovy:
```diff
@@ -498,6 +498,13 @@ class UtilsTest {
     @Test
+    void testToQuotedString() {
+        assert "'a','b','c'" == Utils.toQuotedList(['a', 'b', 'c'])
+        assert "'abc'" == Utils.toQuotedList(['abc'])
```
Add another case:

    assert "''" == Utils.toQuotedList([''])

I'm wondering if `Utils.toQuotedList([])` should be a empty string or not? I think it should be to distinguish it from the case above. That means changing the last test and changing the method also.

    assert "" == Utils.toQuotedList([])

danieldickison replied:
Yeah, I can see this being ok either way (is the method always expected to return a quoted string or is an empty string acceptable?) but having the empty string return value does work better for treating the output as the contents of a Ruby array literal so I'll go with that.

brunobowden commented:
These should be my final comments.

For the `2` method, try deleting the gradle cache:
https://github.com/j2objc-contrib/j2objc-gradle/blob/master/FAQ.md#start-here-for-debugging-aka-its-not-working-aka-dont-panic

Add copyright grant:
https://github.com/j2objc-contrib/j2objc-gradle/blob/master/CONTRIBUTING.md#conditions-for-accepting-pull-requests

Collapse the commits:
https://github.com/j2objc-contrib/j2objc-gradle/blob/master/CONTRIBUTING.md#preparing-your-pull-request-for-submission

brunobowden commented:
Finally, what should happen for `xcodeDebugConfigurations = []`, i.e. no configurations at all? Should that give an error? Will Cocoapods accept it? Try that out and add a comment explaining whatever logic you suggest.

danieldickison commented:
Re: `xcodeDebugConfigurations = []`

If the Podfile contains `:configurations => []`, the pod dependency gets applied to _all_ configurations. This would be unexpected, so I propose that the entire pod line is excluded when the array is empty.

On the other hand, CocoaPods reports an error when `:configurations => ['']`, which I think is fine since, in fact, an empty string is an invalid configuration name.

brunobowden commented:
I feel that `toQuotedList([])` should be an empty string.

It makes sense that for an empty list, the Podfile line is removed
entirely. Add that to the FAQ and the comment in J2objcConfig.

On Wed, Jan 20, 2016 at 9:11 PM Daniel Dickison <notifications@github.com>
wrote:

> Re: xcodeDebugConfigurations = []
>
> If the Podfile contains :configurations => [], the pod dependency gets
> applied to *all* configurations. This would be unexpected, so I propose
> that the entire pod line is excluded when the array is empty.
>
> On the other hand, CocoaPods reports an error when :configurations => [''],
> which I think is fine since, in fact, an empty string is an invalid
> configuration name.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/j2objc-contrib/j2objc-gradle/pull/590#issuecomment-173454535>
> .
>

danieldickison commented:
Ok I think I incorporated all the comments. I went ahead and deleted the "2" suffix from those method names even though I wasn't able to solve the mystery groovy/gradle error locally (I tried deleting the various cache directories, etc, to no avail. Very strange).

danieldickison commented:
Looks like the AppVeyor failure is the same groovy error I'm seeing locally. I'll put the "2" workaround back in there for now, but I'm at a loss as to why these particular method declarations fail.

brunobowden commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/Utils.groovy:
```diff
@@ -511,6 +511,11 @@ class Utils {
 
+    static String toQuotedList(List<String> listString) {
+        if (listString.isEmpty()) return ''
+        else return "'${listString.join("','")}'"
```
We've adopted the style of separating conditional logic over multiple lines:

    if (listString.isEmpty()) {
        return ''
    } else {
        return "'${listString.join("','")}'"
    }

brunobowden commented on src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/XcodeTaskTest.groovy:
```diff
@@ -1107,6 +1112,43 @@ class XcodeTaskTest {
+                "end"]
+        assert expectedLines == podMethodLines
+
+        // Empty configurations list should omit pod line, while an empty string should be included
```
This should be separate tests:

1. testUpdatePodMethods_customConfigurations
1. testUpdatePodMethods_emptyConfigurations
1. testUpdatePodMethods_emptyStringConfigurations

For the last case with `['']`, I think that should be an error. Is there any way it can be valid for a Podfile? I certainly think Xcode wouldn't allow `''` as a valid configuration name.

brunobowden replied:
If it's an error, then you should throw an exception and then catch it in test. See this example elsewhere in the file:

    expectedException.expect(InvalidUserDataException.class)

brunobowden commented on src/main/groovy/com/github/j2objccontrib/j2objcgradle/J2objcConfig.groovy:
```diff
@@ -725,6 +725,31 @@ class J2objcConfig {
+
+    /**
+     * The Xcode build configurations which should link to the generated release libraries.
+     * <p/>
```
Add comment:

    If set to an empty array, then the Debug configuration will be omitted from the "pod method"

danieldickison replied:
Made these changes, though I skipped adding a check for empty config names. That seems like a contrived test case and I think having CocoaPods fail on it is good enough if someone were to try using an empty name.

brunobowden commented:
Given the flakiness, it makes sense to retain the `2`. Very strange and
probably fixed in a later version of Gradle.

On Wed, Jan 20, 2016 at 11:37 PM Daniel Dickison <notifications@github.com>
wrote:

> Looks like the AppVeyor failure is the same groovy error I'm seeing
> locally. I'll put the "2" workaround back in there for now, but I'm at a
> loss as to why these particular method declarations fail.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/j2objc-contrib/j2objc-gradle/pull/590#issuecomment-173484259>
> .
>

brunobowden commented:
@danieldickison - thanks for a great change with good code. I'm happy to have merged this in. We'll probably work on some other fixes for J2ObjC 1.0 before spinning a new version.

danieldickison commented:
Sounds good. Thanks for the code reviews!
